### PR TITLE
Thailand 🇹🇭 Capacities/Pricing/Production(total)

### PR DESCRIPTION
**Pull request by PPsyrius**
tl;dr: Capacity, Price, Production and Consumption data are now in - though switching the parser to the WebSocket version would greatly reduce response time for Production data as well as allow for the Forecasted Production data to be included.

## At a glance:
- The [**Metropolitan Electricity Authority(MEA)/การไฟฟ้านครหลวง (กฟน.)**](https://www.mea.or.th/en/) is responsible for providing electricity in Bangkok, Samut Prakarn, and Nonthaburi provinces while the [**Provincial Electricity Authority (PEA)/การไฟฟ้าส่วนภูมิภาค (กฟภ.)**](https://www.pea.co.th/en) is responsible for the other 73 provinces.
- Thankfully, the electricity unit price in Thailand is the same for the whole country as long as it belongs to the same category with two pricing modes: **Normal** and **Time of Use (TOU)**. The electricity cost was last updated back in [November 2018](https://www.mea.or.th/en/profile/109/111)
- [**Electricity Generating Authority of Thailand (EGAT)/การไฟฟ้าฝ่ายผลิตแห่งประเทศไทย (กฟฝ.)**](https://www.egat.co.th/en/index.php) is the Thai authority responsible for generating electricity, as well as buying/selling generated electricity from/to sub-contractors and neighbouring countries - so we will be focusing mostly on this one.

## Real-time Data
EGAT has give us a real-time system generation, along with forecast data [**here**](https://www.sothailand.com/sysgen/egat/), updating at 2-min interval

I've succeeded in connecting via WebSocket to `wss://www.sothailand.org/genws/sysgen1-server`, but failed to get the return value working so far. If someone wanted to try their hands for better granularity then please give it a try here.

As a side note, unless real-time data categorized by type is available - the carbon intensity rate will likely be very biased because [EGAT tends to have a very high power generation capacity reserve (>40%)](https://www.bangkokpost.com/business/1971303/egat-reining-in-power-reserves) compared to the industry standard practice (~15%).

## Capacity
I will be citing the Thai version of the page instead of the English one here as it's more detailed.

For this, EGAT has two sources: [its own generating capacity](https://www.egat.co.th/index.php?option=com_content&view=article&id=452&Itemid=116) and [private/international sources](https://www.egat.co.th/index.php?option=com_content&view=article&id=355&Itemid=116) - last updated August 2021.

From the data above, here's the translated, summarized data divided by types:
```c++
"biomass": 996.94,
- 352.60 (SPP Firm, Biomass)
- 644.34 (SPP Non-Firm, Biomass)
"coal": 6067.50,
- 2220.00 (EGAT, Lignite Coal)
- 1473.00 (Laos, Lignite Coal)
- 2006.50 (IPP, Bituminus Coal)
- 360.00 (SPP Firm, "Coal")
- 8.00 (SPP Non-Firm, "Coal")
"gas": 27356.00,
- 8262.00 (EGAT, Gas)
- 713.00 (IPP, Gas)
- 10714.00 (IPP, Gas/Diesel)
- 1440.00 (IPP, Gas/Kerosene)
- 5957.00 (SPP Firm, Gas)
- 225.00 (SPP Non-Firm, Gas)
- 45.00 (SPP Non-Firm, Gas/"Coal"/Kerosene)
"geothermal": 0.30,
- 0.30 (EGAT, Geothermal)
"hydro": 6993.73,
- 2972.40 (EGAT, Big Hydro)
- 51.73 (EGAT, Small Hydro)
- 3947.60 (Laos, Hydro)
- 22.00 (SPP Non-Firm, Hydro)
"hydro storage": 1000.00,
- 1000.00 (EGAT, Pump-Storaged Hydro)
"oil": 1497.40,
- 30.40 (EGAT, Diesel)
- 1467.00 (EGAT, Kerosene)
"solar": 442.80,
- 6.80 (EGAT, Solar)
- 436.00 (SPP Non-Firm, Solar)
"wind": 1447.69
- 26.69 (EGAT, Wind)
- 1421.00 (SPP Non-Firm, Wind)
```
In Export/Import Capacity:
```c++
"mixed": 330,
- 30.00 (Malaysia, Malaysia–Thailand HVAC Interconnection Bukit Ketri Station)
- 300.00 (Malaysia, Malaysia–Thailand HVDC Interconnection Gurun Station)
```

## Electricity Price
For this, we can go for the currently-in-use, highest tier pricing for residential users, which is **4.4217 THB/kWh**; last updated in [November 2018](https://www.mea.or.th/en/profile/109/111)

### CODE REVIEW

**Comment by pierresegonne** [comment]
Nice one! Will review on Thursday :)

File: config/co2eq_parameters_lifecycle.json [hunk 1]
```diff
@@ -6576,6 +6576,13 @@
             "value": 199.07165644246462
           }
         ]
+      },
+	  "TH": {
+        "unknown": {
+          "_url": "https://www.iea.org/fuels-and-technologies/electricity",
+          "source": "IEA 2020; assumes 9.1% biomass, 19.3% coal, 64.7% gas, 2.5% hydro, 2.6% solar, 1.7% wind, 0.1% other",
+          "value": 498.9322479978950
```
Comment (pierresegonne) [inline config/co2eq_parameters_lifecycle.json:6584]: I don't get the same exact value:

https://docs.google.com/spreadsheets/d/16j5BVc4L34BDqVE0GeDBv0P3zQupT6zU77Kj5UK7GjE/edit?usp=sharing

Can you additionally please add the factor for direct co2eq as well and link both to the spreadsheet above as the source please? thanks

**Review by pierresegonne — CHANGES_REQUESTED** [review]
You also need to add 

```
"parsers": {
      "price": "TH.fetch_price",
      "production": "TH.fetch_production"
    }
```

to the zone `TH` in `zones.json`.

Other than that and the comment about the emission factor value, we're good to good, very good job :)

File: config/co2eq_parameters_lifecycle.json (hunk 1 above)
Comment (PPsyrius) [reply config/co2eq_parameters_lifecycle.json:6584]: For context, the full breakdown of the number I used for my CO2 lifecycle calculation is in `co2eq_parameter_all.json`
```json
 "TH": {
        "_source": "IEA 2020 (used because Our World in Data has a lot of 'other renewables' generation not broken down by type). Biomass includes waste.",
        "_url": "https://www.iea.org/data-and-statistics/data-browser?country=THAILAND&fuel=Energy%20supply&indicator=ElecGenByFuel",
        "powerOriginRatios": {
          "value": {
            "biomass": 0.0912401506077485,
            "coal": 0.1934589830217160,
            "gas": 0.6467002186068250,
            "geothermal": 0.0000053711750520,
            "hydro": 0.0248631693155512,
            "oil": 0.0007036239318076,
            "solar": 0.0264691506560890,
            "wind": 0.0165593326852115
          }
        }
      },
```
Whereas the 0.1% `other` are actually `geothermal` and `oil` here

Comment (PPsyrius) [reply config/co2eq_parameters_lifecycle.json:6584]: NVM, seems like my calculation is indeed wrong. I will update the number accordingly in this case. Thanks for the double check :)

File: parsers/TH.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
+import re
+
+import arrow
+from requests import get
+from bs4 import BeautifulSoup
+
+EGAT_GENERATION = "https://energy.go.th/index.html"
+#EGAT_GENERATION2 = "https://www.sothailand.com/sysgen/egat/"
```
Comment (pierresegonne) [inline parsers/TH.py:8]: Remove commented line

File: config/co2eq_parameters_lifecycle.json [hunk 3]
```diff
@@ -6576,6 +6576,13 @@
             "value": 199.07165644246462
           }
         ]
+      },
+	  "TH": {
```
Comment (pierresegonne) [inline config/co2eq_parameters_lifecycle.json:6580]: Thanks for correcting the number, can you please link the spreadsheet that I linked before instead please? And add the override emission factor for co2eq_parameters_direct.json

Thanks

**Review by pierresegonne — CHANGES_REQUESTED** [review]

File: config/co2eq_parameters_lifecycle.json (hunk 3 above)
Comment (PPsyrius) [reply config/co2eq_parameters_lifecycle.json:6580]: Added in now with the latest commit

**Review by pierresegonne — APPROVED** [review]
Very nice work! Thanks a lot :)
